core-bundle/src/Routing/Candidates/PageCandidates.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing\Candidates;
+
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Symfony\Component\HttpFoundation\Request;
+
+class PageCandidates extends AbstractCandidates
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var PageRegistry
+     */
+    private $pageRegistry;
+
+    /**
+     * @var bool
+     */
+    private $initialized = false;
+
+    public function __construct(Connection $connection, PageRegistry $pageRegistry)
+    {
+        parent::__construct([], []);
+
+        $this->connection = $connection;
+        $this->pageRegistry = $pageRegistry;
+    }
+
+    public function getCandidates(Request $request): array
+    {
+        $this->initialize();
+
+        $candidates = parent::getCandidates($request);
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->select('id')->from('tl_page');
+
+        $hasRoot = $this->addRootQuery($candidates, $qb, $request->getHttpHost());
+        $hasRegex = $this->addRegexQuery($qb, $request->getPathInfo());
+
+        if ($hasRoot || $hasRegex) {
+            return array_unique(array_merge($candidates, $qb->execute()->fetchAll(FetchMode::COLUMN)));
+        }
+
+        return $candidates;
+    }
+
+    private function addRootQuery(array &$candidates, QueryBuilder $queryBuilder, string $httpHost): bool
+    {
+        if (!\in_array('index', $candidates, true)) {
+            return false;
+        }
+
+        $candidates[] = '/';
+
+        $queryBuilder->orWhere("type='root' AND (dns=:httpHost OR dns='')");
+        $queryBuilder->setParameter('httpHost', $httpHost);
+
+        return true;
+    }
+
+    private function addRegexQuery(QueryBuilder $queryBuilder, string $pathInfo): bool
+    {
+        $pathMap = $this->pageRegistry->getPathRegex();
+
+        if (empty($pathMap)) {
+            return false;
+        }
+
+        $paths = [];
+
+        foreach ($pathMap as $type => $pathRegex) {
+            $paths[] = '(?P<'.$type.'>'.substr($pathRegex, 2, strrpos($pathRegex, '$') - 2).')';
+        }
+
+        preg_match_all(
+            '#^(/'.implode('|/', array_map('preg_quote', $this->urlPrefixes)).')('.implode('|', $paths).')('.implode('|', array_map('preg_quote', $this->urlSuffixes)).')'.'$#sD',
```

Comment:
```suggestion
            '(^/('.implode('|', array_map('preg_quote', $this->urlPrefixes)).')('.implode('|', $paths).')('.implode('|', array_map('preg_quote', $this->urlSuffixes)).')$)sD',
```

Reply:
This is actually not correct because the prefix can be empty, but also a bug in the current implementation. Fixed in 64b388e099414d7449ef24b9ecf0adcd3e4e8297
PS: I'm using `#` simply because that's what Symfony uses.

Reply:
Yes, but `()` is what 🦊es use. I only know it from @ausi as well, though. 😆

Reply:
Not sure if you mean I need to change something? The current regex looks similar to this: `#^(/foo|bar|)(/alias/[a-z]+)(.html|.foo|)$#sD`. If there is an empty option in the prefix or suffix array, it will simply generate an empty or-condition.